### PR TITLE
Add stopword support to product searches

### LINKED ISSUE #19061: CRUD search - support for stopwords
Not sure if its a bug or why was it changed, but since the release of v3.3 product search in admin page hasn't been working like earlier. Now you have to know the exact order of words, or exact order of letters to get correct results. If you have a product called "pink carkey holder" and search  for "pink holder", there will be no results. If you search for "key hol" there will be. Earlier you could search without precise word or letter order and it was alot easier. I have thousands of products with complicated names and the change makes it quite alot harder to search.

### PR DESCRIPTION
**Pull request by mikejolley**
This tweaks the data store search logic to split up terms with spaces, and to remove stop words from searches. 

Code is adapted from WP Query (which does not make any of these methods public sadly) https://github.com/WordPress/WordPress/blob/8a884f562a652bb8f5abe220574f34a3ce1d4126/wp-includes/class-wp-query.php#L1297-L1369

Testing is simple; use dummy data and search for “Beanie with Logo” then “Beanie Logo”.

Before the patch the 2nd search won’t return a result because it tries to find the phrase.

Closes #19061

### CODE REVIEW

File: includes/data-stores/class-wc-data-store-wp.php [hunk 1]
```diff
@@ -430,4 +430,61 @@ public function parse_date_for_wp_query( $query_var, $key, $wp_query_args = arra
...
+
+	/**
+	 * Retrieve stopwords used when parsing search terms.
+	 *
+	 * @since 3.4.0
+	 * @return array Stopwords.
+	 */
+	protected function get_search_stopwords() {
+		// Translators: This is a comma-separated list of very common words that should be excluded from a search, like a, an, and the. These are usually called "stopwords". You should not simply translate these individual words into your language. Instead, look for and provide commonly accepted stopwords in your language.
+		$stopwords = array_map( 'wc_strtolower', array_map( 'trim', explode(
```
Comment (claudiosanches) [inline includes/data-stores/class-wc-data-store-wp.php]: Maybe we could copy this method:

https://developer.wordpress.org/reference/classes/wp_query/get_search_stopwords/

Even the filter, to allow people use the same stopwords.

**Review by claudiosanches — APPROVED** [review]
Looks good to me, but we could copy the WP filter too.

**Comment by claudiulodro** [comment]
Would it be difficult to shoehorn in support for [comma-separated search terms](https://woocommerce.wordpress.com/2018/01/30/woocommerce-3-3-has-been-released/#comment-4365) into this (or a separate PR)? I've confirmed 3.2.x supports them and 3.3.x doesn't.

**Comment by mikejolley** [comment]
@claudiulodro It splits by spaces, so this negates the need for commas?

I don't know why it was supported. Commas seems to be a bit of an antipattern. What if the content you're searching for (phrase) uses commas for example.

**Comment by claudiulodro** [comment]
That's a good point. 

The comma worked as an OR (or something like that), and in 3.2 you can search "wp-pennant, woo-polo" and get both the pennant and the polo shirt. That doesn't seem to be possible any more (even with spaces).  I don't think there is a good way to implement it without making problems in searches for phrases that contain a comma unless we make a full search engine in there.

File: includes/data-stores/class-wc-product-data-store-cpt.php [hunk 2]
```diff
@@ -1326,13 +1326,54 @@ public function get_products( $args = array() ) {
...
 
-		$like_term     = '%' . $wpdb->esc_like( $term ) . '%';
 		$post_types    = $include_variations ? array( 'product', 'product_variation' ) : array( 'product' );
 		$post_statuses = current_user_can( 'edit_private_products' ) ? array( 'private', 'publish' ) : array( 'publish' );
 		$type_join     = '';
 		$type_where    = '';
 		$status_where  = '';
 
+		// See if search term contains OR keywords.
+		if ( stristr( $term, ' OR ' ) ) {
```
Comment (kloon) [inline includes/data-stores/class-wc-product-data-store-cpt.php]: Should this be uppercase only? If you test with lowercase it does not work.

File: includes/data-stores/class-wc-product-data-store-cpt.php [hunk 3]
```diff
@@ -1326,13 +1326,54 @@ public function get_products( $args = array() ) {
...
+				$search_terms = array( $term_group );
+			}
+
+			$term_group_query = '';
+			$searchand        = '';
+
+			foreach ( $search_terms as $search_term ) {
+				$like              = '%' . $wpdb->esc_like( $search_term ) . '%';
+				$term_group_query .= $wpdb->prepare( " {$searchand} ( ( posts.post_title LIKE %s) OR ( posts.post_content LIKE %s ) OR ( postmeta.meta_key = '_sku' AND postmeta.meta_value LIKE %s ) )", $like, $like, $like ); // @codingStandardsIgnoreLine.
+				$searchand         = ' AND ';
```
Comment (kloon) [inline includes/data-stores/class-wc-product-data-store-cpt.php]: Should this be AND? I tested with + , and in both cases it used the AND. My understanding would be that using , should OR this? IE. Searching for `hoodie+pocket,hoodie+logo` should return both the hoodie with logo and hoodie with pocket products, but instead with this it returns nothing?

File: includes/data-stores/class-wc-product-data-store-cpt.php (hunk 2 above)
Comment (mikejolley) [reply includes/data-stores/class-wc-product-data-store-cpt.php]: Should be lower. Fixing.

File: includes/data-stores/class-wc-product-data-store-cpt.php (hunk 3 above)
Comment (mikejolley) [reply includes/data-stores/class-wc-product-data-store-cpt.php]: hoodie+pocket or hoodie+logo

**Comment by kloon** [comment]
Good to merge!

**Pull request merged**